cloudevents/client.go

```diff
@@ -0,0 +1,131 @@
+package cloudevents
+
+import (
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+// Client wraps Builder, and is intended to be configured for a single event
+// type and target
+type Client struct {
+	Builder
+	Target string
+
+	send chan interface{}
+	done chan bool
+}
+
+const (
+	// MAX_SEND_CHANNEL is the max buffered channel length for the send channel
+	MAX_SEND_CHANNEL = 10
+)
+
+// NewClient returns a CloudEvent Client used to send CloudEvents. It is
+// intended that a user would create a new client for each tuple of eventType
+// and target. This is a n optional helper method to avoid the tricky creation
+// of the embedded Builder struct.
+func NewClient(eventType, source, target string) *Client {
+	c := &Client{
+		Builder: Builder{
+			Source:    source,
+			EventType: eventType,
+		},
+		Target: target,
+	}
+	return c
+}
+
+// Send creates a request based on the client's settings and sends the data
+// struct to the target set for this client. It returns error if there was an
+// issue sending the event, otherwise nil means the event was accepted.
+func (c *Client) Send(data interface{}) error {
+	resp, err := c.RequestSend(data)
+	if err != nil {
+		return err
+	}
+	if accepted(resp) {
+		return nil
+	}
+	return fmt.Errorf("error sending cloudevent: %s", status(resp))
+}
+
+// RequestSend uses the internal builder to make a request with the provided
+// data struct using the previously set parameters of event type and target.
+// Use this instead of client.Send() if processing of the http response directly
+// is required.
+func (c *Client) RequestSend(data interface{}) (*http.Response, error) {
+	req, err := c.Build(c.Target, data)
+	if err != nil {
+		return nil, err
+	}
+
+	client := &http.Client{}
+	return client.Do(req)
+}
+
+// Channel returns a channel that can be used to invoke Client.Send via a chan.
+// This method has a side effect of another thread to monitor the send channel.
+// Call client.Close() to shutdown the monitor thread.
+// Experimental, error handling is not fully developed.
+func (c *Client) Channel() chan<- interface{} {
```

Comment:
Is this part of the public interface? This seems like it should be private, since sharing around a `Client` could cause random panics if someone decides to `close(c.Channel())`.

Reply:
I was playing with this idea:

```
func main() {
	c := cloudevents.NewClient(
		"dev.knative.cloudevent.example",
		"https://github.com/knative/pkg#cloudevents-example",
		"http://localhost:8080",
	)
	ch := c.Channel()

	data := &Example{
		Message: "hello, world!",
	}

	for i := 0; i < 10; i++ {
		data.Sequence = i
		ch <- *data
	}

	time.Sleep(1 * time.Second)
	c.Done()
}
```

Reply:
I am going to pull it for now, I can always add a client channel wrapper